### DISCUSSION TITLE
Support for multiple schemas in a multi-tenant graphql service?

### DISCUSSION THREAD

**Discussion by polivenok** [Netflix/dgs-framework#888] [Q&A]
Hello,

I am curious is it possible to have multiple Graphql schemas and switch between them, and what would be your suggestions to achieve that? The use case that can benefit from this is following: multi tenant application where each tenant (client) can have their own schema built using [dynamic schema](https://netflix.github.io/dgs/advanced/dynamic-schemas/).

Technically it would mean that for each request we can have client/tenant identifier and I would like to use schema specific to this tenant/client. I am aware of `ReloadSchemaIndicator.reloadSchema` method, so this can be overwritten and always return `true`, but this would not be very performant, as schema will be built on each request. As an option, I can try to cache dynamic schema building as much as possible, but maybe there are some other options.

Thanks,
Dmytro

**Comment by GrantGochnauer** [comment]
Hi DGS Team - 

I'm working with @polivenok and to build on his initial question, consider the following scenario:

* A multi-tenant GraphQL API
* Each tenant has a combination of a statically defined schema and a dynamically generated one (based on their configuration). This is working great based on the dynamic schema features.
* A JWT is passed into the service that includes a tenant identifier

When each request comes into the API, we want to be able to fetch the already computed schema for that tenant based on their tenantId. 

All of the above is working great with the exception that it doesn't appear that we have a way to hook into DGS to intercept which schema should be served based on the tenant identifier without asking for a full schema reload. We are already caching our dynamic schema generation but when we tell DGS/GraphQL to reload the schema because a new tenantID request came in, the performance hit is anywhere between 200ms-500ms. 

Ideally we'd have a way to hook into `DgsSchemaProvider` or `DefaultDgsQueryExecutor` such that we could cache a fully computed schema based on tenant and then give that back to DGS. 

We'd love for your guidance on this one.
Thank you in advance,
Grant

**Reply by srinivasankavitha** [reply]
Thanks for the detailed context. We haven't come across multi-tenant usages so far, so this is a first. I have a few questions for my clarification. Can what you are describing be achieved by providing different graphql endpoints for each of the tenants? Then the requests would be automatically processed for the  endpoint it is targeting.

**Reply by GrantGochnauer** [reply]
Hi @srinivasankavitha -

Sure thing. Our system is effectively a document management system and each customer configures the types of documents and the fields and field types as part of a standard implementation. We want to create a unified public GraphQL api for all tenants and have the schema be strongly typed based on their system configuration.

The closest example I can give you to what we are trying to achieve is this:
https://www.contentful.com/developers/docs/references/graphql/#/reference/schema-generation

The different URL strategy is interesting - do you know if we can expose different schemas at different endpoints within the same DGS application?

/graphql/tenant1/
/graphql/tenant2/
(or some kind of standard pattern above)

Where each of these are specific to the customer and then we don't have to worry about switching the "current" schema and reload it on the thread?

Thanks

**Comment by srinivasankavitha** [comment]
We don't currently support it, but might be a good way to support your use
case compared to trying to get the schema provider knowledgeable about
tenant Ids. @Bernardo Gomez Palacio ***@***.***> @Paul Bakker
***@***.***> - thoughts about this approach?

**Reply by GrantGochnauer** [reply]
@srinivasankavitha Thoughts on providing a hook for users to intercept the fetching of schemas? This would enable us to provide our own "multi-tenant" implementation on the fly.

**Reply by srinivasankavitha** [reply]
@GrantGochnauer - that does sound like a useful feature to support. However, we don't have the bandwidth currently to prioritize this feature since it is not something we use internally yet. We do welcome PR contributions though, so if you have ideas on what you would like to see, feel free to open a PR we can iterate on.

**Reply by merrillse** [reply]
We have the exact same "multi-tenant" use case as outlined in this thread. Have there been any updates to this? (We are looking for something similar to GraphQLSchemaServletProvider in the deprecated graphql-java-servlet project which allowed for a cached schema to be provided per client. Or, could the GraphqlFieldVisibility feature of graphql-java (https://www.graphql-java.com/documentation/field-visibility/) be a possible solution with a hook provided by DGS?)

**Reply by GrantGochnauer** [reply]
We ended up writing a very custom solution that combines static and dynamic graphql schemas - every request into the system will dynamically load the schema specific for the request/tenant (based on the JWT value).

**Comment by srinivasankavitha** [comment]
Nice find! I agree, I think the GraphqlFieldVisibility feature could be
easily supported in the framework to address this scenario as you suggest.
We could add @DgsFieldVisibility as a new annotation for defining custom field
visibility and the framework would register when processing the schema.

Currently, you can do the same by adding that directly in a `DgsCodeRegistry` : https://netflix.github.io/dgs/advanced/dynamic-schemas/#creating-datafetchers-programmatically

I think that might actually work for you (with the disclaimer that we haven't actually explored this scenario yet for our internal use cases.)

**Reply by merrillse** [reply]
Thank you very much GrantGochnauer and srinivasankavitha for your responses! We will investigate the DgsCodeRegistry solution. Grant, (after a few days of agonizing over how to do this) it would be very interesting to hear about your approach, if there was a fair and equitable way to share that information. Thanks again!

**Reply by GrantGochnauer** [reply]
@merrillse Sure - a quick summary of our approach with a few code snippets - there is a lot going on so may be hard to parse through all of this.

* Implemented a DgsComponent called `DynamicTypeDefinitionRegistry` that includes the `@DgsTypeDefinitionRegistry` annotation that calls our own service that generates the schema. We also implement a DgsTypeResolver
```
  /**
   * We make a call from this DgsTypeDefinitionRegistry to our own service implementation so that we can wrap it with Spring's caching
   */
  @DgsTypeDefinitionRegistry
  public TypeDefinitionRegistry generateSchemaTypes() {
    return dynamicSchemaGeneration.generateSchemaTypes();
  }

 @DgsTypeResolver(name = DgsConstants.DOCUMENTTYPE.TYPE_NAME)
  public String resolveDocumentTypes(com.vodori.pepper.cloud.content.sdk.graphql.types.DocumentType document) {
}
```

The service method looks something like this:
```
 /**
     * The DgsTypeDefinitionRegistry is where we can register GraphQL types in addition to what we have defined in the schema file.
     * This method retrieves the DocumentTypes for the current tenant and uses that information to create tenant specific GraphQL types.
     * This method is responsible for registering all Documents, Filters, Input types, and Unions for a tenant.
     */
    @Cacheable(value = "dgsSchemasByTenant", keyGenerator = CacheConstants.ENVIRONMENT_ID_CACHE_KEY_GENERATOR)
    public TypeDefinitionRegistry generateSchemaTypes() {
        TypeDefinitionRegistry typeDefinitionRegistry = new TypeDefinitionRegistry();
        Optional<String> tenantIdFromSecurityContext = JwtUtil.getTenantIdFromSecurityContext();

        if(tenantIdFromSecurityContext.isEmpty()) {
            return typeDefinitionRegistry;
        }

       typeDefinitionRegistry.add(customDocument);
       return typeDefinitionRegistry;

```


Then, we implement a DgsCodeRegistry to register dynamic document data fetchers. Such as:
```
 @DgsCodeRegistry
    public GraphQLCodeRegistry.Builder registerDynamicDocumentDataFetchers(GraphQLCodeRegistry.Builder codeRegistryBuilder, TypeDefinitionRegistry registry) {

        var collectionTypes = registry.getAllImplementationsOf(
                        InterfaceTypeDefinition.newInterfaceTypeDefinition().name(DgsConstants.DOCUMENTTYPE.TYPE_NAME).build()).stream()
                .filter(implementingTypeDefinition -> !implementingTypeDefinition.getName().equals(DgsConstants.DOCUMENT.TYPE_NAME)).toList();

        collectionTypes.forEach(type -> {
            FieldCoordinates coordinates = FieldCoordinates.coordinates(DgsConstants.QUERY_TYPE, Introspector.decapitalize(type.getName()));
            DataFetcher<DataFetcherResult<Document>> df = dfe -> {
                DocumentFilter documentFilter = modelMapper.map(dfe.getArgument("where"), DocumentFilter.class);
                List<String> requestedFields = SelectionSetUtility.getMongoFieldsFromSelectionSet(dfe.getSelectionSet());
                GraphQLObjectType graphQLObjectType = (GraphQLObjectType) dfe.getFieldType();

                List<DocumentType> allMergedDocumentTypes = documentTypeService.getAllMergedDocumentTypes();

                Map<String,String> mapOfDocumentTypes =
                        allMergedDocumentTypes.stream().collect(Collectors.toMap(
                                documentType -> generateDocumentTypeName(documentType.getConfigKey()), DocumentType::getConfigKey));

                String kind = mapOfDocumentTypes.get(graphQLObjectType.getName());
                final com.vodori.cloud.data.mongo.documents.model.Document doc = documentService.findDocumentByKindAndDocumentFilter(kind, documentFilter, requestedFields);

                if (doc == null) {
                    return null;
                } else {
                    Document document = modelMapper.map(doc, Document.class);

                    Map<String, com.vodori.cloud.data.mongo.documents.model.Document> map = Map.ofEntries(
                            new AbstractMap.SimpleEntry<>(documentUtility.createDocumentKey(doc), doc));

                    return DataFetcherResult.<Document>newResult().data(document).localContext(Map.of("documentsMap", map)).build();
                }

            };
            codeRegistryBuilder.dataFetcher(coordinates, df);
        });

        return codeRegistryBuilder;
```

Lastly, we want to refresh the schema on every request (since each request could be a different tenant)
```
/**
 * This class implements the ReloadSchemaIndicator and allows us to use our own custom logic to decide when the schema should be reloaded.
 * For now we always want the schema reloaded.
 */
@DgsComponent
public class ReloadSchema implements DefaultDgsQueryExecutor.ReloadSchemaIndicator {

  /**
   * We always want DGS to think it needs to reload the schema because each request coming in, we need to grab the correct schema for the
   * tenant based on the JWT tenantId value
   */
  @Override
  public boolean reloadSchema() {
    return true;
  }
}
```

**Reply by highcohesion** [reply]
Thank you very much Grant! This has been enormously helpful for us.

**Reply by etooley** [reply]
Hi @GrantGochnauer, I'm working with Steven. Do you still have the 200-500ms delay with this code you are showing? We are getting a similar result now.

**Reply by GrantGochnauer** [reply]
@etooley We are not - but we're using  caching quite heavily. We also realized that a significant performance issue was due to the use of ModelMapper (http://modelmapper.org/) and instead moved to https://mapstruct.org/ which is orders of magnitude faster. 

We are using Spring's caching annotation with a custom tenant-aware key generator
```
@Cacheable(value = "dgsSchemasByTenant", keyGenerator = CacheConstants.ENVIRONMENT_ID_CACHE_KEY_GENERATOR)
    public TypeDefinitionRegistry generateSchemaTypes() { }
```

We are also caching the underlying service calls to get all the tenant-specific types that we use to generate the schema avoid DB calls when generating schemas.
